csi/controller.go

```diff
@@ -817,15 +821,184 @@ func (s *OsdCsiServer) DeleteSnapshot(
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to connect to SDK server: %v", err)
+	}
+
+	// Get secret if any was passed
+	ctx = s.setupContextWithToken(ctx, req.GetSecrets())
+	ctx, cancel := grpcutil.WithDefaultTimeout(ctx)
+	defer cancel()
+	volumes := api.NewOpenStorageVolumeClient(conn)
+
+	sdkRequest := &api.SdkVolumeInspectRequest{
+		VolumeId: req.GetSnapshotId(),
+	}
+
+	// Snapshots are treated as volume, inspect for this volume.
+	resp, err := volumes.Inspect(ctx, sdkRequest)
+	if err != nil {
+		status, ok := status.FromError(err)
+		if ok && status.Code() == codes.NotFound {
+			return &csi.ListSnapshotsResponse{
+				Entries: []*csi.ListSnapshotsResponse_Entry{},
+			}, nil
+		}
+
+		return nil, err
+	}
+
+	snap := csi.Snapshot{
+		SizeBytes:      int64(resp.Volume.GetSpec().GetSize()),
+		SnapshotId:     snapshotId,
+		SourceVolumeId: resp.Volume.GetSource().Parent,
+		CreationTime:   resp.Volume.Ctime,
+		ReadyToUse:     true,
+	}
+
+	return &csi.ListSnapshotsResponse{
+		Entries: []*csi.ListSnapshotsResponse_Entry{
+			{
+				Snapshot: &snap,
+			},
+		},
+	}, nil
+}
+
+// listMultipleSnapshots gets all snapshots for a given
+// volumeId or all snapshots if no volumeId is provided.
+// This involves calling both InspectWithFilters and SnapshotEnumerateWithFilters
+// as SnapshotEnumerateWithFilters only returns all snapshotIds.
+// TODO: optimize this function if a SnapshotInspectWithFilters with pagination is implemented.
+func (s *OsdCsiServer) listMultipleSnapshots(
+	ctx context.Context,
+	req *csi.ListSnapshotsRequest,
+) (*csi.ListSnapshotsResponse, error) {
+	sourceVolumeId := req.GetSourceVolumeId()
+	startingToken := req.GetStartingToken()
+	maxEntries := req.GetMaxEntries()
+
+	logrus.Debugf("ListSnapshots for multiple snapshots received. sourceVolumeId: %s, startingToken: %s, maxEntries: %v",
+		sourceVolumeId,
+		startingToken,
+		maxEntries,
+	)
+
+	// Get grpc connection
+	conn, err := s.getConn()
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to connect to SDK server: %v", err)
+	}
+
+	// Get secret if any was passed
+	ctx = s.setupContextWithToken(ctx, req.GetSecrets())
+	ctx, cancel := grpcutil.WithDefaultTimeout(ctx)
+	defer cancel()
+	volumes := api.NewOpenStorageVolumeClient(conn)
+
+	// InspectWithFilters for all volumes
+	volumesResp, err := volumes.InspectWithFilters(ctx, &api.SdkVolumeInspectWithFiltersRequest{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Unable to get all volumes: %v", err)
+	}
+
+	// Get all SnapshotIDs. Filter by source ID if provided.
+	snapshotsReq := &api.SdkVolumeSnapshotEnumerateWithFiltersRequest{}
+	if sourceVolumeId != "" {
+		snapshotsReq.VolumeId = sourceVolumeId
+	}
+	snapshotsResp, err := volumes.SnapshotEnumerateWithFilters(ctx, snapshotsReq)
+	if err != nil {
+		errStatus, ok := status.FromError(err)
+		if ok && errStatus.Code() == codes.NotFound {
+			return &csi.ListSnapshotsResponse{}, nil
```

Comment:
This this code is used to check if the `sourceVolumeId` exists, it should run before running line 929. Line 929 could take some time to get since it is getting all volumes

Reply:
Makes sense, good idea